pennylane/transforms/split_non_commuting.py

```diff
@@ -0,0 +1,111 @@
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the tape transform that splits non-commuting terms
+"""
+# pylint: disable=protected-access
+import pennylane as qml
+import numpy as np
+
+
+def split_non_commuting(tape):
+    r"""
+    Splits a tape measuring non-commuting observables into groups of commuting observables.
+
+    Args:
+        tape (.QuantumTape): the tape used when calculating the expectation value
+            of the Hamiltonian
+
+    Returns:
+        tuple[list[.QuantumTape], function]: Returns a tuple containing a list of
+        quantum tapes to be evaluated, and a function to be applied to these
+        tape executions to restore the ordering of the inputs.
+
+    **Example**
+
+    We can create a tape with non-commuting observables:
+
+    .. code-block:: python3
+
+        import pennylane as qml
+        with qml.tape.QuantumTape() as tape:
+            qml.expval(qml.PauliZ(0))
+            qml.expval(qml.PauliY(0))
+        tapes, processing_fn = qml.transforms.split_non_commuting(tape)
+
+    Now ``tapes`` is a list of two tapes, each for one of the non-commuting terms:
+
+    >>> [t.observables for t in tapes]
+    [[expval(PauliZ(wires=[0]))], [expval(PauliY(wires=[0]))]]
+
+    The processing function becomes important when creating the commuting groups as the order of the inputs has been modified:
+
+    .. code-block:: python3
+
+        with qml.tape.QuantumTape() as tape:
+            qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+            qml.expval(qml.PauliX(0) @ qml.PauliX(1))
+            qml.expval(qml.PauliZ(0))
+            qml.expval(qml.PauliX(0))
+        tapes, processing_fn = qml.transforms.split_non_commuting(tape)
+
+    In this example, the groupings are ``group_coeffs = [[0,2], [1,3]]`` and processing_fn makes sure that the final output is of the same shape and ordering:
+
+    >>> processing_fn(tapes)
+    tensor([tensor(expval(PauliZ(wires=[0]) @ PauliZ(wires=[1])), dtype=object, requires_grad=True),
+        tensor(expval(PauliX(wires=[0]) @ PauliX(wires=[1])), dtype=object, requires_grad=True),
+        tensor(expval(PauliZ(wires=[0])), dtype=object, requires_grad=True),
+        tensor(expval(PauliX(wires=[0])), dtype=object, requires_grad=True)],
+       dtype=object, requires_grad=True)
+
+    """
+
+    obs_fn = {qml.measurements.Expectation: qml.expval,
+              qml.measurements.Variance: qml.var,
+              qml.measurements.Sample: qml.sample}
```

Comment:
Remember that we have probabilities, state, and density matrices.

Reply:
Are there scenarios where you return multiple probabilities, states and density matrices in a quantum function?

Reply:
> I wonder if we need to extend group_observables to work on MeasurementProcess objects as well. Then we can simply use apply(m) instead of having to manage obs_fn[return_type](o). apply(m) seems much more robust to future changes and less generally fragile and specific.

I agree @albi3ro! This would be a nice improvement.

> Are there scenarios where you return multiple probabilities, states and density matrices in a quantum function?

@Qottmann it comes up occasionally depending on research needs, but it is not common. The most common workflows tend to be:

- One expval
- multiple expvals
- One probability

Reply:
Indeed something like the following would be possible at least for probs:
```python
import pennylane as qml
dev = qml.device("default.qubit", wires=2)

@qml.qnode(dev)
def circuit():
    qml.Hadamard(0)
    qml.T(0)
    return qml.probs(op=qml.PauliX(wires=0)), qml.probs(op=qml.PauliZ(wires=0))
circuit()
```

Currently outputs an error `WireError: Did not find some of the wires (PauliX(wires=[0]),) on device with wires (0, 1).` (Having added `qml.measurements.Probability: qml.probs` in the obs_fn list). But changing the wire of one of the operators to 1 produces the normal, expected behavior. When wires are collected in qml.probs, are those "popped", i.e. taken "out" of the operator that is given?